mlir/lib/Gradient/IR/GradientOps.cpp

```diff
@@ -106,13 +142,180 @@ LogicalResult GradOp::verify()
+}
+
+//===----------------------------------------------------------------------===//
+// VJPOp, CallOpInterface
+//===----------------------------------------------------------------------===//
+
+CallInterfaceCallable VJPOp::getCallableForCallee() { return getCalleeAttr(); }
+
+Operation::operand_range VJPOp::getArgOperands() { return getOperands(); }
+
+//===----------------------------------------------------------------------===//
+// VJPOp, SymbolUserOpInterface
+//===----------------------------------------------------------------------===//
+
+LogicalResult VJPOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+    // Check that the callee attribute refers to a valid function.
+    auto callee = ({
+         auto cattr = this->getCalleeAttr();
+         func::FuncOp fn = symbolTable.lookupNearestSymbolFrom<func::FuncOp>(
+             this->getOperation(), cattr);
+         if (!fn)
+             return this->emitOpError("invalid function name specified: ") << cattr;
+         fn;
+    });
+
+    // Check gradient input parameters
+    auto r1 = ::verifyGradInputs(
+        this, callee,
+        this->getCalleeOperands(),
+        GradOp::compDiffArgIndices(this->getDiffArgIndices()));
+    if (r1.failed()) {
+        return r1;
+    }
+
+    auto calleeResultTypes = callee.getFunctionType().getResults();
+
+    auto cotTypes = ({
+        std::vector<Type> out;
+        auto cotangOperands = OperandRange(
+            this->operand_begin() + callee.getFunctionType().getNumInputs(),
+            this->operand_end());
+        for (auto c : cotangOperands) {
+            out.push_back(c.getType());
+        }
+        out;
+    });
+
+    // Check that callee results have the same size as cotangent inputs
+    if (calleeResultTypes.size() != cotTypes.size()) {
+        return this->emitOpError(
+            "number of callee results does not match the number of cotangent arguments")
+               << " expected " << cotTypes.size()
+               << " but got " << calleeResultTypes.size();
+    }
+
+    // Check that callee results have the same types as cotangent inputs
+    for (size_t i=0; i<cotTypes.size(); i++) {
+        auto cotType = cotTypes[i];
+        auto crType = calleeResultTypes[i];
+        if (cotType != crType) {
+            return this->emitOpError("callee result type does not match the cotangent type")
+               << " callee result " << i
+               << " was expected to be of type " << cotType
+               << " but got " << crType;
+        }
+    }
+
+    return success();
+}
+
+//===----------------------------------------------------------------------===//
+// VJPOp Extra methods
+//===----------------------------------------------------------------------===//
+
+std::vector<mlir::Value> VJPOp::getCalleeOperands()
+{
+    return std::vector<mlir::Value>(
+        this->operand_begin(),
+        this->operand_begin() + (this->getNumOperands() - this->getCalleeResults().size()));
+}
```

Comment:
Why not use the `op.getParams()` function generated by ODS?

Reply:
Oh, you have revealed a serious problem but this. We actually can't use `SameVariadicOperandsSize` for JVP because it doesn't hold for non-full argnums.

Reply:
Should be fixed now

Reply:
Yes, I switched to the AttrSizedOperandSegments already, and added the check.